### ISSUE TITLE
Excessive string copies and allocation in DbKVS queries

### ISSUE THREAD

**Issue by schlosna** [palantir/atlasdb#735]
While looking through some profiles generated from simple DbKVS gets on https://github.com/palantir/atlasdb/tree/ds/atlasdb-perf-jmh I noticed a ton of memory allocations and CPU time being spent in `PostgresQueryFactory.wrapQueryWithIncludeValue` and various `SQLString` methods doing query generation and parsing (on the order of several hundred MB/s).

```
Stack Trace                                                                                                                                                                TLABs         Total TLAB Size(bytes)  Pressure(%)
java.util.Arrays.copyOf(char[], int)                                                                                                                                       1,256         3,603,624,912           39.47
   java.lang.AbstractStringBuilder.ensureCapacityInternal(int)                                                                                                             1,256         3,603,624,912           39.47
      java.lang.AbstractStringBuilder.append(String)                                                                                                                       1,235         3,568,896,720           39.089
         java.lang.StringBuilder.append(String)                                                                                                                            1,235         3,568,896,720           39.089
            com.palantir.atlasdb.keyvalue.dbkvs.impl.postgres.PostgresQueryFactory.wrapQueryWithIncludeValue(String, String, boolean)                                      289           939,836,880             10.294
               com.palantir.atlasdb.keyvalue.dbkvs.impl.postgres.PostgresQueryFactory.getLatestCellQuery(Cell, long, boolean)                                              289           939,836,880             10.294
                  com.palantir.atlasdb.keyvalue.dbkvs.impl.AbstractDbReadTable.getLatestCells(Map, boolean)                                                                289           939,836,880             10.294
                     com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs$4.apply(DbReadTable)                                                                                   289           939,836,880             10.294
            com.palantir.nexus.db.sql.BasicSQLString$FinalSQLString.toString()                                                                                             235           769,486,816             8.428
               com.palantir.nexus.db.sql.BasicSQL.runCancellablyInternal(PreparedStatement, BasicSQL$ResultSetVisitor, BasicSQLString$FinalSQLString, BasicSQL$AutoClose)  235           769,486,816             8.428
            com.palantir.util.TextUtils.replaceAllWhitespace(String, String)                                                                                               217           709,131,360             7.767
               com.palantir.util.TextUtils.removeAllWhitespace(String)                                                                                                     217           709,131,360             7.767
                  com.palantir.nexus.db.sql.SQLString.getUnregisteredQuery(String)                                                                                         217           709,131,360             7.767
                     com.palantir.nexus.db.sql.SqlConnectionHelper.selectLightResultSetUnregisteredQuery(Connection, String, Object[])                                     217           709,131,360             7.767
                        com.palantir.nexus.db.sql.ConnectionBackedSqlConnectionImpl.selectLightResultSetUnregisteredQuery(String, Object[])                                217           709,131,360             7.767
            com.palantir.nexus.db.sql.SQLString.<init>(String)                                                                                                             86            288,934,720             3.165
               com.palantir.nexus.db.sql.SQLString.getUnregisteredQuery(String)                                                                                            86            288,934,720             3.165
                  com.palantir.nexus.db.sql.SqlConnectionHelper.selectLightResultSetUnregisteredQuery(Connection, String, Object[])                                        86            288,934,720             3.165
                     com.palantir.nexus.db.sql.ConnectionBackedSqlConnectionImpl.selectLightResultSetUnregisteredQuery(String, Object[])                                   86            288,934,720             3.165
                        com.palantir.atlasdb.keyvalue.dbkvs.impl.BatchedDbReadTable.run(FullQuery)                                                                         86            288,934,720             3.165
                           com.palantir.atlasdb.keyvalue.dbkvs.impl.AbstractDbReadTable.getLatestCells(Map, boolean)                                                       86            288,934,720             3.165
                              com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs$4.apply(DbReadTable)                                                                          86            288,934,720             3.165
                                 com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs$4.apply(Object)                                                                            86            288,934,720             3.165
                                    com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs.runRead(TableReference, Function)                                                       86            288,934,720             3.165
            com.palantir.atlasdb.keyvalue.dbkvs.impl.postgres.PostgresQueryFactory.getLatestCellQuery(Cell, long, boolean)                                                 83            282,104,360             3.09
               com.palantir.atlasdb.keyvalue.dbkvs.impl.AbstractDbReadTable.getLatestCells(Map, boolean)                                                                   83            282,104,360             3.09
                  com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs$4.apply(DbReadTable)                                                                                      83            282,104,360             3.09
            java.lang.StringBuilder.append(Object)                                                                                                                         80            268,565,784             2.942
            com.palantir.nexus.db.sql.SQLString.<init>(String, String, DBType)                                                                                             59            188,996,480             2.07
            com.palantir.nexus.db.sql.BasicSQLString.toString()                                                                                                            40            36,746,248              0.402
```

```
Stack Trace                                                                        TLABs         Total TLAB Size(bytes)  Pressure(%)
java.lang.String.substring(int, int)                                               226           749,437,840             84.594
   java.lang.String.subSequence(int, int)                                          116           388,454,824             43.848
      java.util.regex.Pattern.split(CharSequence, int)                             116           388,454,824             43.848
         java.lang.String.split(String, int)                                       116           388,454,824             43.848
            java.lang.String.split(String)                                         116           388,454,824             43.848
               com.palantir.nexus.db.sql.SQLString.canonicalizeString(String)      116           388,454,824             43.848
                  com.palantir.nexus.db.sql.SQLString.getUnregisteredQuery(String) 116           388,454,824             43.848
```

```
Stack Trace                                                                                                                              Sample Count  Percentage(%)
java.util.regex.Pattern$Curly.match(Matcher, int, CharSequence)                                                                          28            16.092
org.apache.commons.lang.text.StrBuilder.ensureCapacity(int)                                                                              25            14.368
   org.apache.commons.lang.text.StrBuilder.append(String)                                                                                25            14.368
      org.apache.commons.lang.StringUtils.join(Object[], String, int, int)                                                               15            8.621
         org.apache.commons.lang.StringUtils.join(Object[], String)                                                                      15            8.621
            com.palantir.nexus.db.sql.SQLString.canonicalizeString(String)                                                               15            8.621
               com.palantir.nexus.db.sql.SQLString.getUnregisteredQuery(String)                                                          15            8.621
                  com.palantir.nexus.db.sql.SqlConnectionHelper.selectLightResultSetUnregisteredQuery(Connection, String, Object[])      15            8.621
                     com.palantir.nexus.db.sql.ConnectionBackedSqlConnectionImpl.selectLightResultSetUnregisteredQuery(String, Object[]) 15            8.621
                        com.palantir.atlasdb.keyvalue.dbkvs.impl.BatchedDbReadTable.run(FullQuery)                                       15            8.621
                           com.palantir.atlasdb.keyvalue.dbkvs.impl.AbstractDbReadTable.getLatestCells(Map, boolean)                     15            8.621
                              com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs$4.apply(DbReadTable)                                        15            8.621
                                 com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs$4.apply(Object)                                          15            8.621
      org.apache.commons.lang.text.StrBuilder.append(Object)                                                                             10            5.747
```

```
Stack Trace                                                                                                                                        Sample Count  Percentage(%)
java.util.regex.Pattern$Curly.match(Matcher, int, CharSequence)                                                                                    79            10.064
   java.util.regex.Pattern$Start.match(Matcher, int, CharSequence)                                                                                 79            10.064
      java.util.regex.Matcher.search(int)                                                                                                          79            10.064
         java.util.regex.Matcher.find()                                                                                                            79            10.064
            java.util.regex.Pattern.split(CharSequence, int)                                                                                       78            9.936
               java.lang.String.split(String, int)                                                                                                 78            9.936
                  java.lang.String.split(String)                                                                                                   78            9.936
                     com.palantir.nexus.db.sql.SQLString.canonicalizeString(String)                                                                78            9.936
                        com.palantir.nexus.db.sql.SQLString.getUnregisteredQuery(String)                                                           78            9.936
                           com.palantir.nexus.db.sql.SqlConnectionHelper.selectLightResultSetUnregisteredQuery(Connection, String, Object[])       78            9.936
                              com.palantir.nexus.db.sql.ConnectionBackedSqlConnectionImpl.selectLightResultSetUnregisteredQuery(String, Object[])  78            9.936
                                 com.palantir.atlasdb.keyvalue.dbkvs.impl.BatchedDbReadTable.run(FullQuery)                                        78            9.936
                                    com.palantir.atlasdb.keyvalue.dbkvs.impl.AbstractDbReadTable.getLatestCells(Map, boolean)                      78            9.936
                                       com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs$4.apply(DbReadTable)                                         78            9.936
                                          com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs$4.apply(Object)                                           78            9.936
                                             com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs.runRead(TableReference, Function)                      78            9.936
                                                com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs.get(TableReference, Map)                            78            9.936
                                                   com.palantir.atlasdb.keyvalue.impl.ForwardingKeyValueService.get(TableReference, Map)           78            9.936
```

**Comment by schlosna** [comment]
Some additional notes from a [single constant get benchmark](https://github.com/palantir/atlasdb/commit/5d5ac235113924dd46930eeccea0b1d510cb40ae#diff-437808c674991f014a16a56dab22434aR108):

```
Iteration   1: n = 52736, mean = 758 us/op, p{0.00, 0.50, 0.90, 0.95, 0.99, 0.999, 0.9999, 1.00} = 269, 489, 910, 1145, 1890, 151064, 169320, 235667 us/op
                 ·gc.alloc.rate:               232.098 MB/sec
                 ·gc.alloc.rate.norm:          46170.690 B/op
                 ·gc.churn.PS_Eden_Space:      217.874 MB/sec
                 ·gc.churn.PS_Eden_Space.norm: 43341.088 B/op
                 ·gc.count:                    2.000 counts
                 ·gc.time:                     400.000 ms
                 ·stack:                       <delayed till summary>

Iteration   2: n = 52097, mean = 767 us/op, p{0.00, 0.50, 0.90, 0.95, 0.99, 0.999, 0.9999, 1.00} = 265, 490, 900, 1124, 1835, 139671, 154707, 155451 us/op
                 ·gc.alloc.rate:                   229.287 MB/sec
                 ·gc.alloc.rate.norm:              46177.948 B/op
                 ·gc.churn.PS_Eden_Space:          294.779 MB/sec
                 ·gc.churn.PS_Eden_Space.norm:     59367.788 B/op
                 ·gc.churn.PS_Survivor_Space:      1.274 MB/sec
                 ·gc.churn.PS_Survivor_Space.norm: 256.590 B/op
                 ·gc.count:                        3.000 counts
                 ·gc.time:                         404.000 ms
                 ·stack:                           <delayed till summary>

Iteration   3: n = 46068, mean = 869 us/op, p{0.00, 0.50, 0.90, 0.95, 0.99, 0.999, 0.9999, 1.00} = 281, 502, 933, 1180, 2025, 151781, 152830, 153092 us/op
                 ·gc.alloc.rate:                   202.455 MB/sec
                 ·gc.alloc.rate.norm:              46177.670 B/op
                 ·gc.churn.PS_Eden_Space:          189.381 MB/sec
                 ·gc.churn.PS_Eden_Space.norm:     43195.669 B/op
                 ·gc.churn.PS_Survivor_Space:      0.233 MB/sec
                 ·gc.churn.PS_Survivor_Space.norm: 53.163 B/op
                 ·gc.count:                        2.000 counts
                 ·gc.time:                         259.000 ms
                 ·stack:                           <delayed till summary>

Iteration   4: n = 44812, mean = 892 us/op, p{0.00, 0.50, 0.90, 0.95, 0.99, 0.999, 0.9999, 1.00} = 281, 555, 1151, 1493, 2519, 146257, 153879, 153879 us/op
                 ·gc.alloc.rate:                   197.218 MB/sec
                 ·gc.alloc.rate.norm:              46178.902 B/op
                 ·gc.churn.PS_Eden_Space:          188.199 MB/sec
                 ·gc.churn.PS_Eden_Space.norm:     44066.957 B/op
                 ·gc.churn.PS_Survivor_Space:      0.078 MB/sec
                 ·gc.churn.PS_Survivor_Space.norm: 18.278 B/op
                 ·gc.count:                        2.000 counts
                 ·gc.time:                         264.000 ms
                 ·stack:                           <delayed till summary>

Iteration   5: 14:55:40.720 [com.palantir.atlasdb.performance.benchmarks.KvsPutBenchmarks.singleGet-jmh-worker-2] WARN  c.p.d.c.logging.FileLogCollector - docker containers were still running when log collection stopped
n = 45495, mean = 878 us/op, p{0.00, 0.50, 0.90, 0.95, 0.99, 0.999, 0.9999, 1.00} = 285, 488, 908, 1149, 1964, 152044, 154521, 155451 us/op
                 ·gc.alloc.rate:                   133.667 MB/sec
                 ·gc.alloc.rate.norm:              40311.335 B/op
                 ·gc.churn.PS_Eden_Space:          144.003 MB/sec
                 ·gc.churn.PS_Eden_Space.norm:     43428.528 B/op
                 ·gc.churn.PS_Survivor_Space:      0.246 MB/sec
                 ·gc.churn.PS_Survivor_Space.norm: 74.177 B/op
                 ·gc.count:                        2.000 counts
                 ·gc.time:                         254.000 ms
                 ·stack:                           <delayed till summary>

Result "singleGet":
  N = 241208
  mean =    828.837 ±(99.9%) 38.416 us/op

  Histogram, us/op:
    [     0.000,  25000.000) = 240848
    [ 25000.000,  50000.000) = 8
    [ 50000.000,  75000.000) = 4
    [ 75000.000, 100000.000) = 0
    [100000.000, 125000.000) = 4
    [125000.000, 150000.000) = 48
    [150000.000, 175000.000) = 292
    [175000.000, 200000.000) = 0
    [200000.000, 225000.000) = 0
    [225000.000, 250000.000) = 4
    [250000.000, 275000.000) = 0

  Percentiles, us/op:
      p(0.0000) =    265.216 us/op
     p(50.0000) =    502.784 us/op
     p(90.0000) =    951.296 us/op
     p(95.0000) =   1218.560 us/op
     p(99.0000) =   2084.864 us/op
     p(99.9000) = 151519.232 us/op
     p(99.9900) = 154402.816 us/op
     p(99.9990) = 235297.285 us/op
     p(99.9999) = 235667.456 us/op
    p(100.0000) = 235667.456 us/op

Secondary result "·gc.alloc.rate":
  198.945 ±(99.9%) 152.773 MB/sec [Average]
  (min, avg, max) = (133.667, 198.945, 232.098), stdev = 39.675
  CI (99.9%): [46.172, 351.718] (assumes normal distribution)

Secondary result "·gc.alloc.rate.norm":
  45003.309 ±(99.9%) 10099.836 B/op [Average]
  (min, avg, max) = (40311.335, 45003.309, 46178.902), stdev = 2622.895
  CI (99.9%): [34903.473, 55103.145] (assumes normal distribution)

Secondary result "·gc.churn.PS_Eden_Space":
  206.847 ±(99.9%) 214.876 MB/sec [Average]
  (min, avg, max) = (144.003, 206.847, 294.779), stdev = 55.803
  CI (99.9%): [≈ 0, 421.723] (assumes normal distribution)

Secondary result "·gc.churn.PS_Eden_Space.norm":
  46680.006 ±(99.9%) 27341.531 B/op [Average]
  (min, avg, max) = (43195.669, 46680.006, 59367.788), stdev = 7100.509
  CI (99.9%): [19338.475, 74021.538] (assumes normal distribution)

Secondary result "·gc.churn.PS_Survivor_Space":
  0.366 ±(99.9%) 1.995 MB/sec [Average]
  (min, avg, max) = (≈ 0, 0.366, 1.274), stdev = 0.518
  CI (99.9%): [≈ 0, 2.361] (assumes normal distribution)

Secondary result "·gc.churn.PS_Survivor_Space.norm":
  80.442 ±(99.9%) 395.262 B/op [Average]
  (min, avg, max) = (≈ 0, 80.442, 256.590), stdev = 102.648
  CI (99.9%): [≈ 0, 475.703] (assumes normal distribution)

Secondary result "·gc.count":
  11.000 ±(99.9%) 0.001 counts [Sum]
  (min, avg, max) = (2.000, 2.200, 3.000), stdev = 0.447
  CI (99.9%): [11.000, 11.000] (assumes normal distribution)

Secondary result "·gc.time":
  1581.000 ±(99.9%) 0.001 ms [Sum]
  (min, avg, max) = (254.000, 316.200, 404.000), stdev = 78.417
  CI (99.9%): [1581.000, 1581.000] (assumes normal distribution)
```

**Comment by clockfort** [comment]
Woke up too early and looked into this because I was bored.

I locally refactored PostgresQueryFactory.getLatestCellQuery to create a string builder explicitly for the query, gave it an initial capacity of 1024, and passed this stringbuilder to the formatter helper method for it to use as well.
This resulted in a 35% speedup (w/ 2% variance) in my tests. 

StringBuilder, at least in my version of java, pre-sizes to only 16 characters and doubles the capacity whenever it runs out, but this results in a bunch of extra work, since the minimum size for something that calls this query is >700 length, and possibly can grow considerably larger.

**Comment by clockfort** [comment]
I do find it semi-disappointing that the compiler is smart enough to turn repeated string appends into a string builder, but not smart enough to make at least slightly better stringbuilder buffer allocation. 

Someone else was semi-dissapointed about it on the openjdk list in 2014 ("StringBuilder buffer allocation support in compiler"), but it looks like nothing came about from it.

**Comment by schlosna** [comment]
I think its worthwhile to just presize the StringBuilder ourselves and push a single one through all the relevant methods to minimize the expansions.

Yeah, there have been a few threads on OpenJDK lists about optimizations that could be done in javac as well as the JIT.

* http://markmail.org/message/xp4ff65kbkayeqoh
* http://markmail.org/message/mrhy6ystcirswhqf

**Comment by gbrova** [comment]
@clockfort is there a branch / PR that we should look into merging?

**Comment by gsheasby** [comment]
Fixed by #2470.
